src/layout.rs

```diff
@@ -782,4 +764,409 @@ mod tests {
         const _HORIZONTAL_LAYOUT: Layout = Layout::new().horizontal_margin(1);
         const _VERTICAL_LAYOUT: Layout = Layout::new().vertical_margin(1);
     }
+
+    /// Tests for the `Layout::split()` function.
+    ///
+    /// There are many tests in this as the number of edge cases that are caused by the interaction
+    /// between the constraints is quite large. The tests are split into sections based on the type
+    /// of constraints that are used.
+    ///
+    /// These tests are characterization tests. This means that they are testing the way the code
+    /// currently works, and not the way it should work. This is because the current behavior is not
+    /// well defined, and it is not clear what the correct behavior should be. This means that if
+    /// the behavior changes, these tests should be updated to match the new behavior.
+    mod layout_split {
+        use pretty_assertions::assert_eq;
+
+        use crate::{
+            prelude::{Constraint::*, *},
+            widgets::{Paragraph, Widget},
+        };
+
+        /// Test that the given constraints applied to the given area result in the expected
+        /// layout. Each chunk is filled with a letter repeated as many times as the width of the
+        /// chunk. The resulting buffer is compared to the expected string.
+        #[track_caller]
+        fn test(area: Rect, constraints: &[Constraint], expected: &str) {
+            let layout = Layout::default()
+                .direction(Direction::Horizontal)
+                .constraints(constraints)
+                .split(area);
+            let mut buffer = Buffer::empty(area);
+            for (i, c) in ('a'..='z').take(constraints.len()).enumerate() {
+                let s: String = c.to_string().repeat(area.width as usize);
+                Paragraph::new(s).render(layout[i], &mut buffer);
+            }
+            assert_eq!(buffer.content, Buffer::with_lines(vec![expected]).content);
+        }
+
+        #[test]
+        fn length() {
+            test(Rect::new(0, 0, 1, 1), &[Length(0)], "a"); // underflow
+            test(Rect::new(0, 0, 1, 1), &[Length(1)], "a"); // exact
+            test(Rect::new(0, 0, 1, 1), &[Length(2)], "a"); // overflow
```

Comment:
does `overflow` in this case mean that it allocates more than the given area, or that the individual elements themself should be longer?

Reply:
Mostly that the allocated space is more than the available space. Overflow is not strictly the correct term, but I couldn't think of a better one. I wanted to convey simply that there was a rationale for each test being there.

Reply:
maybe add terminology along the lines of

```rust
/// underflow = less than given space used
/// exact = space is used up
/// overflow = constraint requires more but got less
```
at least that is my current understanding of the given terminology